### PR TITLE
[sailfish-components-webview] Add C++ WebEngine API. Contributes to JB#36227

### PR DESCRIPTION
**Pull request by rainemak**

### CODE REVIEW

File: import/plugin.cpp [hunk 1]
```diff
@@ -84,52 +43,31 @@ void SailfishOSWebViewPlugin::initializeEngine(QQmlEngine *engine, const char *u
...
+    // Disable low-precision-buffer so that background underdraw works
+    // correctly.
+    engineSettings->setPreference("layers.low-precision-buffer", QVariant::fromValue<bool>(false));
+
+    // Don't expose any protocol handlers by default and don't warn about those.
+    engineSettings->setPreference(QStringLiteral("network.protocol-handler.external-default"), false);
+    engineSettings->setPreference(QStringLiteral("network.protocol-handler.expose-all"), false);
+    engineSettings->setPreference(QStringLiteral("network.protocol-handler.warn-external-default"), false);
+
+    engine->rootContext()->setContextProperty("MozContext", webEngine);
```
Comment (chriadam) [inline import/plugin.cpp]: I guess we can't change the name of the context property at this point, yet.  Maybe add two: MozContext and WebEngine, then in "new" code (e.g. in this repo) use the WebEngine property?  Slowly migrate existing code (in Sailfish Browser) over to the WebEngine property at our leisure?

File: lib/webengine.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+#ifndef SAILFISHOS_WEBENGINE_H
+#define SAILFISHOS_WEBENGINE_H
+
+#include <QString>
+#include <qmozcontext.h>
+
+namespace SailfishOS {
+
+class WebEngine : public QMozContext {
```
Comment (chriadam) [inline lib/webengine.h]: Are we trying to get rid of QMozContext altogether at some stage?  Or not?  Just wondering if it makes sense to extend the public WebEngine class here, or instead provide some API and internally call through to the appropriate function of a QMozContext private data instance...

This is probably fine as-is, just thought I'd ask.

File: lib/webenginesettings.cpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+#include <QtCore/QSize>
+#include <QtCore/QtMath>
+#include <QtGui/QGuiApplication>
+#include <QtGui/QScreen>
+
+Q_GLOBAL_STATIC(SailfishOS::WebEngineSettings, webEngineSettingsInstance)
+
+#define SAILFISH_WEBENGINE_DEFAULT_PIXEL_RATIO 1.5
+
+static bool testScreenDimensions(qreal pixelRatio) {
```
Comment (chriadam) [inline lib/webenginesettings.cpp]: prefer brace on separate line

File: lib/webenginesettings.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    QScreen *screen = QGuiApplication::primaryScreen();
+    qreal w = screen->size().width() / pixelRatio;
+    qreal h = screen->size().height() / pixelRatio;
+    return fmod(w, 1.0) == 0 && fmod(h, 1.0) == 0;
+}
+
+const QSettings &quickSettings()
+{
+    static const QSettings settings(QSettings::SystemScope, QStringLiteral("QtProject"), QStringLiteral("QtQuick2"));
```
Comment (chriadam) [inline lib/webenginesettings.cpp]: Why like this?  I would have expected UserScope with an applicationName derived from the application binary path (or read from /proc/pid etc)

**Comment by chriadam** [comment]
A couple of minor comments / discussion points, but LGTM.

File: import/webengine/plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    Q_OBJECT
+    Q_PLUGIN_METADATA(IID "Sailfish.WebEngine" )
+
+public:
+    void initializeEngine(QQmlEngine *, const char *)
+    {
+        // TODO : How to deal with custom default UA. We have also means to customize
+        // site specific UA overrides.
+        SailfishOS::WebEngine::initialize(QStandardPaths::writableLocation(QStandardPaths::CacheLocation));
+        SailfishOS::WebEngineSettings::initialize();
```
Comment (rainemak) [inline import/webengine/plugin.cpp]: Let's figure out once we have more use cases that how should we actually do the initialization. Maybe this is something that should not happen here at all rather we could have explicit QML API. On the other hand, this kind of approach would be quite convenient.

File: import/plugin.cpp (hunk 1 above)
Comment (chriadam) [inline import/plugin.cpp]: If WebEngine is a QML singleton, the engine will own it (and delete it on engine dtor) which could cause issues if you try to use a Q_GLOBAL_STATIC instance.  But yes I agree with your plan, just implementation details we need to figure out.

**Comment by chriadam** [comment]
LGTM

**Comment by chriadam** [comment]
Other thing we need going forward (but shouldn't block this PR) is settings to enable or disable JavaScript and Image Loading

File: import/plugin.cpp (hunk 1 above)
Comment (rainemak) [inline import/plugin.cpp]: We can already create instances of WebEngine/WebEngineSettings that will share private implementation. Problem that I'm currently seeing is commented in https://github.com/sailfishos/sailfish-components-webview/pull/7#pullrequestreview-697748 e.i. how to deal initialization?

Maybe need to just rid off initialization arguments and handle custom cases separately.

**Pull request merged**